lib/db/structs.go

```diff
@@ -196,98 +195,297 @@ func (vl VersionList) String() string {
-				vl = vl.insertAt(i, nv)
-				return vl, removedFV, removedAt, i, nil
-			}
+	for ; i < len(vl.RawVersions); i++ {
+		// Insert our new version
+		vl.RawVersions, added, err = checkInsertAt(vl.RawVersions, i, folder, device, []byte(file.FileName()), file.FileVersion(), file.IsInvalid(), file.IsDeleted(), file, true, t)
+		if err != nil {
+			return vl, FileVersion{}, FileVersion{}, FileVersion{}, false, false, false, err
+		}
+		if added {
+			break
 		}
 	}
+	if i == len(vl.RawVersions) {
+		// Append to the end
+		vl.RawVersions = append(vl.RawVersions, newFileVersion(device, file.FileVersion(), file.IsInvalid(), file.IsDeleted()))
+	}
 
-	// We didn't find a position for an insert above, so append to the end.
-	vl.Versions = append(vl.Versions, nv)
+	newFV, _ := vl.GetGlobal() // We just inserted something above, can't be empty
 
-	return vl, removedFV, removedAt, len(vl.Versions) - 1, nil
-}
+	if !haveOldGlobal {
+		return vl, newFV, FileVersion{}, removedFV, false, haveRemoved, true, nil
+	}
+
+	globalChanged := true
+	if oldFV.IsInvalid() == newFV.IsInvalid() && oldFV.Version.Equal(newFV.Version) {
+		globalChanged = false
+	}
 
-func (vl VersionList) insertAt(i int, v FileVersion) VersionList {
-	vl.Versions = append(vl.Versions, FileVersion{})
-	copy(vl.Versions[i+1:], vl.Versions[i:])
-	vl.Versions[i] = v
-	return vl
+	return vl, newFV, oldFV, removedFV, true, haveRemoved, globalChanged, nil
 }
 
 // pop returns the VersionList without the entry for the given device, as well
-// as the removed FileVersion and the position, where that FileVersion was.
-// If there is no FileVersion for the given device, the position is -1.
-func (vl VersionList) pop(device []byte) (VersionList, FileVersion, int) {
-	for i, v := range vl.Versions {
-		if bytes.Equal(v.Device, device) {
-			vl.Versions = append(vl.Versions[:i], vl.Versions[i+1:]...)
-			return vl, v, i
-		}
+// as the removed FileVersion, whether it was found/removed at all and whether
+// the global changed in the process.
+func (vl VersionList) pop(folder, device, name []byte, t readOnlyTransaction) (VersionList, FileVersion, bool, bool, error) {
+	invDevice, i, j, ok := vl.findDevice(device)
+	if !ok {
+		return vl, FileVersion{}, false, false, nil
+	}
+	globalPos := vl.findGlobal()
+
+	if vl.RawVersions[i].deviceCount() == 1 {
+		fv := vl.RawVersions[i]
+		vl.RawVersions = popVersionAt(vl.RawVersions, i)
+		return vl, fv, true, globalPos == i, nil
+	}
+
+	if invDevice {
+		vl.RawVersions[i].InvalidDevices = popDeviceAt(vl.RawVersions[i].InvalidDevices, j)
+	} else {
+		vl.RawVersions[i].Devices = popDeviceAt(vl.RawVersions[i].Devices, j)
 	}
-	return vl, FileVersion{}, -1
+	// If the last valid device of the previous global was removed above,
+	// the next entry is now the global entry (unless all entries are invalid).
+	if len(vl.RawVersions[i].Devices) == 0 && globalPos == i {
+		return vl, vl.RawVersions[i], true, globalPos == vl.findGlobal(), nil
```

Comment:
There's literally no way `globalPos == vl.findGlobal()` can be true here, as we've already established there are no valid devices in that version? (I'm not sure if that means we can just simplify a little or if there's a logic error in the thinking somewhere else...)

Reply:
A version with no valid devices can be the global version, if there's no valid devices at all. That's what I tried to explain in the comment above. If we were sure there's still a valid device, we could just return `true` here.